### DISCUSSION TITLE
Missing HEIC support

### DISCUSSION THREAD

**Discussion by Byter3** [nextcloud/all-in-one#2438] [Questions] [answered]
Fresh install with every optional containers. NC AIO v4.9 NC 26.0.1.
Copied a couple of Gigabytes of HEIC/movs and the HEIC files are not showing any previews evven after installing preview generator. I can see the MOVs mp4s and jpgs, but not the HEIC. In addition to that, in the "photos" app there's simply nothing shown.

**Comment by szaimen** [comment]
Hi, do you have something in the Nextcloud logs?

**Reply by Byter3** [reply]
Hi, I found this in the log, to be more precises Nextcloud's log was flooded with this message.

`[no app in context] Error: GuzzleHttp\Exception\ConnectException: cURL error 28: Operation timed out after 30003 milliseconds with 0 bytes received (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for http://nextcloud-aio-imaginary:9000/pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D at <<closure>>

 0. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Handler/CurlFactory.php line 158
    GuzzleHttp\Handler\CurlFactory::createRejection("*** sensitive parameters replaced ***")
 1. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Handler/CurlFactory.php line 110
    GuzzleHttp\Handler\CurlFactory::finishError(["GuzzleHttp\\Handler\\CurlHandler"], "*** sensitive parameters replaced ***", ["GuzzleHttp\\Handler\\CurlFactory"])
 2. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Handler/CurlHandler.php line 47
    GuzzleHttp\Handler\CurlFactory::finish(["GuzzleHttp\\Handler\\CurlHandler"], "*** sensitive parameters replaced ***", ["GuzzleHttp\\Handler\\CurlFactory"])
 3. /var/www/html/lib/private/Http/Client/DnsPinMiddleware.php line 114
    GuzzleHttp\Handler\CurlHandler->__invoke("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
 4. /var/www/html/3rdparty/guzzlehttp/guzzle/src/PrepareBodyMiddleware.php line 64
    OC\Http\Client\DnsPinMiddleware->OC\Http\Client\{closure}("*** sensitive parameters replaced ***")
 5. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Middleware.php line 31
    GuzzleHttp\PrepareBodyMiddleware->__invoke("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
 6. /var/www/html/3rdparty/guzzlehttp/guzzle/src/RedirectMiddleware.php line 71
    GuzzleHttp\Middleware::GuzzleHttp\{closure}("*** sensitive parameters replaced ***")
 7. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Middleware.php line 63
    GuzzleHttp\RedirectMiddleware->__invoke("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
 8. /var/www/html/3rdparty/guzzlehttp/guzzle/src/HandlerStack.php line 75
    GuzzleHttp\Middleware::GuzzleHttp\{closure}("*** sensitive parameters replaced ***")
 9. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Client.php line 331
    GuzzleHttp\HandlerStack->__invoke("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
10. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Client.php line 168
    GuzzleHttp\Client->transfer("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
11. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Client.php line 187
    GuzzleHttp\Client->requestAsync("*** sensitive parameters replaced ***")
12. /var/www/html/lib/private/Http/Client/Client.php line 297
    GuzzleHttp\Client->request("*** sensitive parameters replaced ***", "http://nextclou ... e", ["/mnt/ncdata/fi ... "])
13. /var/www/html/lib/private/Preview/Imaginary.php line 141
    OC\Http\Client\Client->post("http://nextclou ... e", [["[{\"operation ... ]])
14. /var/www/html/lib/private/Preview/GeneratorHelper.php line 62
    OC\Preview\Imaginary->getCroppedThumbnail(["OC\\Files\\Node\\File"], 256, 256, false)
15. /var/www/html/lib/private/Preview/Generator.php line 274
    OC\Preview\GeneratorHelper->getThumbnail(["OC\\Preview\\Imaginary"], ["OC\\Files\\Node\\File"], 256, 256, false)
16. /var/www/html/lib/private/Preview/Generator.php line 153
    OC\Preview\Generator->getSmallImagePreview(["OC\\Files\\SimpleFS\\SimpleFolder"], ["OC\\Files\\Node\\File"], "image/jpeg", "", false)
17. /var/www/html/lib/private/Preview/Generator.php line 116
    OC\Preview\Generator->generatePreviews(["OC\\Files\\Node\\File"], [[250,250,false,"fill"]], "image/jpeg")
18. /var/www/html/lib/private/PreviewManager.php line 192
    OC\Preview\Generator->getPreview(["OC\\Files\\Node\\File"], 250, 250, false, "fill", null)
19. /var/www/html/core/Controller/PreviewController.php line 144
    OC\PreviewManager->getPreview(["OC\\Files\\Node\\File"], 250, 250, false, "fill")
20. /var/www/html/core/Controller/PreviewController.php line 113
    OC\Core\Controller\PreviewController->fetchPreview(["OC\\Files\\Node\\File"], 250, 250, true, false, "fill")
21. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 230
    OC\Core\Controller\PreviewController->getPreviewByFileId(2761, 250, 250, true, false, "fill")
22. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 137
    OC\AppFramework\Http\Dispatcher->executeController(["OC\\Core\\Cont ... "], "getPreviewByFileId")
23. /var/www/html/lib/private/AppFramework/App.php line 183
    OC\AppFramework\Http\Dispatcher->dispatch(["OC\\Core\\Cont ... "], "getPreviewByFileId")
24. /var/www/html/lib/private/Route/Router.php line 315
    OC\AppFramework\App::main("OC\\Core\\Controller\\PreviewController", "getPreviewByFileId", ["OC\\AppFramewo ... "], ["core.Preview.getPreviewByFileId"])
25. /var/www/html/lib/base.php line 1056
    OC\Route\Router->match("/core/preview")
26. /var/www/html/index.php line 36
    OC::handleRequest()

GET /core/preview?fileId=2761&c=8da425c70a8ea2f8c58043585d6f773a&x=250&y=250&forceIcon=0&a=1
from 192.168.60.1 by AG at 2023-04-28T22:32:47+00:00`

**Comment by szaimen** [comment, marked as answer]
Can you check if this patch fixes the problem for you? https://github.com/nextcloud/server/pull/37148.patch

**Reply by Byter3** [reply]
Is this something that got pushed to the beta channel? I can switch from the :latest, but other then that, I'm not sure how to implement it.

**Reply by szaimen** [reply]
In this case applying the patch could look like this:
```
# Go into the container
sudo docker exec --user www-data -it nextcloud-aio-nextcloud bash
```
```
# Now inside the container
wget https://patch-diff.githubusercontent.com/raw/nextcloud/server/pull/37148.patch
git apply --check 37148.patch && git apply 37148.patch
exit
```

**Reply by Byter3** [reply]
Thank you for the patch and for the guidance, it solved it!

**Comment by radityakurnianto** [comment]
> Fresh install with every optional containers. NC AIO v4.9 NC 26.0.1. Copied a couple of Gigabytes of HEIC/movs and the HEIC files are not showing any previews evven after installing preview generator. I can see the MOVs mp4s and jpgs, but not the HEIC. In addition to that, in the "photos" app there's simply nothing shown.

I also got the same in version 29. 

The file itself is ok, downloadable and can be preview perfectly on windows. Could some body point what sould be done to fix this?

The log said that

```
{
   "reqId": "Qctb69klyg6GnNqIYpNa",
   "level": 3,
   "time": "2024-05-23T13:22:32+00:00",
   "remoteAddr": "100.88.33.103",
   "user": "raditya",
   "app": "core",
   "method": "GET",
   "url": "/index.php/core/preview.png?file=%2FGallery%2FRaditya%2Fff8a22af-6b80-444e-a785-a4005b045c52.HEIC&x=540&y=1086&a=1&mode=cover&forceIcon=0",
   "message": "File: /raditya/files/Gallery/Raditya/ff8a22af-6b80-444e-a785-a4005b045c52.HEIC Imagick says:",
   "userAgent": "Mozilla/5.0 (Android) Nextcloud-android/3.29.0",
   "version": "29.0.0.19",
   "exception": {
      "Exception": "ImagickException",
      "Message": "negative or zero image size `/mnt/SSD_2TB/Media/Raditya/ff8a22af-6b80-444e-a785-a4005b045c52.HEIC' @ error/image.c/CloneImage/794",
      "Code": 425,
      "Trace": [
         {
            "file": "/var/www/html/nextcloud/lib/private/Preview/HEIC.php",
            "line": 119,
            "function": "autoOrient",
            "class": "Imagick",
            "type": "->"
         },
         {
            "file": "/var/www/html/nextcloud/lib/private/Preview/HEIC.php",
            "line": 76,
            "function": "getResizedPreview",
            "class": "OC\\Preview\\HEIC",
            "type": "->"
         },
         {
            "file": "/var/www/html/nextcloud/l...oud/lib/private/AppFramework/Http/Dispatcher.php",
            "line": 138,
            "function": "executeController",
            "class": "OC\\AppFramework\\Http\\Dispatcher",
            "type": "->"
         },
         {
            "file": "/var/www/html/nextcloud/lib/private/AppFramework/App.php",
            "line": 184,
            "function": "dispatch",
            "class": "OC\\AppFramework\\Http\\Dispatcher",
            "type": "->"
         },
         {
            "file": "/var/www/html/nextcloud/lib/private/Route/Router.php",
            "line": 338,
            "function": "main",
            "class": "OC\\AppFramework\\App",
            "type": "::"
         },
         {
            "file": "/var/www/html/nextcloud/lib/base.php",
            "line": 1050,
            "function": "match",
            "class": "OC\\Route\\Router",
            "type": "->"
         },
         {
            "file": "/var/www/html/nextcloud/index.php",
            "line": 49,
            "function": "handleRequest",
            "class": "OC",
            "type": "::"
         }
      ],
      "File": "/var/www/html/nextcloud/lib/private/Preview/HEIC.php",
      "Line": 119,
      "message": "File: /raditya/files/Gallery/Raditya/ff8a22af-6b80-444e-a785-a4005b045c52.HEIC Imagick says:",
      "exception": [],
      "CustomMessage": "File: /raditya/files/Gallery/Raditya/ff8a22af-6b80-444e-a785-a4005b045c52.HEIC Imagick says:"
   },
   "id": "664f45811fdf0"
}


[core] Error: File: /raditya/files/Gallery/Raditya/ff8a22af-6b80-444e-a785-a4005b045c52.HEIC Imagick says: 	GET /index.php/core/preview.png?file=%2FGallery%2FRaditya%2Fff8a22af-6b80-444e-a785-a4005b045c52.HEIC&x=540&y=1086&a=1&mode=cover&forceIcon=0 	from 100.88.33.103 by raditya at 23 May 2024, 20:22:32 ```
